include/deal.II/lac/solver_idr.h

```diff
@@ -436,43 +429,58 @@ SolverIDR<VectorType>::solve(const MatrixType &        A,
             Mk_inv.vmult(gamma, phik);
           }
 
-          {
-            v = r;
+          v = r;
 
-            unsigned int j = 0;
-            for (unsigned int i = k; i < s; ++i, ++j)
-              v.add(-1.0 * gamma(j), G[i]);
-            preconditioner.vmult(vhat, v);
+          if (step > 1)
+            {
+              for (unsigned int i = k, j = 0; i < s; ++i, ++j)
+                v.add(-gamma(j), G[i]);
+            }
+
+          preconditioner.vmult(uhat, v);
 
-            uhat = vhat;
+          if (step > 1)
+            {
+              uhat.sadd(omega, gamma(0), U[k]);
+              for (unsigned int i = k + 1, j = 1; i < s; ++i, ++j)
+                uhat.add(gamma(j), U[i]);
+            }
+          else
             uhat *= omega;
-            j = 0;
-            for (unsigned int i = k; i < s; ++i, ++j)
-              uhat.add(gamma(j), U[i]);
-            A.vmult(ghat, uhat);
-          }
+
+          A.vmult(G[k], uhat);
 
           // Update G and U
-          // Orthogonalize ghat to Q0,..,Q_{k-1}
-          // and update uhat
-          for (unsigned int i = 0; i < k; ++i)
+          // Orthogonalize G[k] to Q0,..,Q_{k-1} and update uhat
+          if (k > 0)
             {
-              double alpha = (Q[i] * ghat) / M(i, i);
-              ghat.add(-alpha, G[i]);
-              uhat.add(-alpha, U[i]);
+              double alpha = Q[0] * G[k] / M(0, 0);
+              for (unsigned int i = 1; i < k; ++i)
+                {
+                  const double alpha_old = alpha;
+                  alpha = G[k].add_and_dot(-alpha, G[i - 1], Q[i]) / M(i, i);
+
+                  // update uhat every other iteration to reduce vector access
+                  if (i % 2 == 1)
+                    uhat.add(-alpha_old, U[i - 1], -alpha, U[i]);
+                }
+              M(k, k) = G[k].add_and_dot(-alpha, G[k - 1], Q[k]);
+              if (k % 2 == 1)
+                uhat.add(-alpha, U[k - 1]);
             }
-          G[k] = ghat;
-          U[k] = uhat;
+          else
+            M(k, k) = G[k] * Q[k];
+
+          U[k].swap(uhat);
 
           // Update kth column of M
-          for (unsigned int i = k; i < s; ++i)
+          for (unsigned int i = k + 1; i < s; ++i)
             M(i, k) = Q[i] * G[k];
 
-          // Orthogonalize r to Q0,...,Qk,
-          // update x
+          // Orthogonalize r to Q0,...,Qk, update x
           {
-            double beta = phi(k) / M(k, k);
-            r.add(-1.0 * beta, G[k]);
+            const double beta = phi(k) / M(k, k);
```

Comment:
`double`?

Reply:
Do you want me to use `VectorType::value_type` like in the other PR? I can do that, if you agree with changing the matrix types and all other stuff currently in `double` variables, such as this one here: https://github.com/dealii/dealii/blob/fb2242de546e203f273a4ab674b7b7de9994c84d/include/deal.II/lac/solver_idr.h#L423-L424
Or are you suggesting something else?

Reply:
> Do you want me to use VectorType::value_type like in the other PR? I can do that, if you agree with changing the matrix types and all other stuff currently in double variables, such as this one here:

Yes, sounds reasonable :+1: